### DISCUSSION TITLE
EXCEPT DISTINCT

### DISCUSSION THREAD

**Discussion by tkawachi** [EvgSkv/logica#80] [Q&A] [answered]
Is it possible to express EXCEPT DISTINCT in Logica?

I would like to verify that the results of a query are as expected, and it seems that Standard SQL will allow me to diff tables with EXCEPT DISTINCT and UNION ALL, although there are some limitations.
ref. https://stackoverflow.com/a/57912743

**Comment by EvgSkv** [comment, marked as answer]
This can be done with disjunction and negation.

```python
A(x) :- x in [1,2,3,4];
B(x) :- x in [2,3,4,5];

Difference(x) :- (A(x), ~B(x)) | (B(x), ~A(x));
```
You can add more info to the result to simplify debugging:

```python
A(x) :- x in [1,2,3,4];
B(x) :- x in [2,3,4,5];

Difference(x, reason:) :- 
  (A(x), ~B(x), reason == "A but not B") |
  (B(x), ~A(x), reason == "B but not A");
```

Note that here the difference is taken for a predicate over 1 argument.
It would work the same way if you repeat a few arguments.

Taking difference of full rows would require some more work.

If the goal is to unit-test the program, it could be more productive to set up a testing framework and test in with a simple text diff.
A nice thing about testing framework is that it would generate golden files automatically after manual approval.
[Here is an example of a Python script doing it.](https://github.com/EvgSkv/logica/blob/main/integration_tests/run_tests.py)
The plan is to make this kind of unittesting easier. Let me know if this could help your work, so we can prioritize it appropriately.

Please let me know if you have further questions.

**Reply by tkawachi** [reply]
Thank you for the quick answer.

As you said, it would be better to use golden files for unit tests. I'll refer to the test script you provided.
